src/mds/MDCache.cc

```diff
@@ -6542,9 +6543,86 @@ void MDCache::start_recovered_truncates()
 }
 
 
+class C_MDS_purge_completed_finish : public MDCacheLogContext {
+  interval_set<inodeno_t> inos;
+  version_t inotablev;
+  LogSegment *ls; 
+public:
+  C_MDS_purge_completed_finish(MDCache *m,
+			       const interval_set<inodeno_t>& i,
+			       version_t iv,
+			       LogSegment *_ls)
+    : MDCacheLogContext(m),
+      inos(i),
+      inotablev(iv),
+      ls(_ls) {}
+  void finish(int r) override {
+    assert(r == 0);
+    if (inotablev) {
+      ls->purge_inodes_finish(inos);      
+      mdcache->mds->inotable->apply_release_ids(inos);
+      assert(mdcache->mds->inotable->get_version() == inotablev);
+    }
+  }
+};
 
+void MDCache::start_purge_inodes(){
+  dout(10) << "start_purge_inodes" << dendl;
+  for (auto p = mds->mdlog->segments.begin();
+       p != mds->mdlog->segments.end();
+       ++p) {
+    LogSegment *ls = p->second;
+    purge_inodes(ls->purge_inodes, ls);
+  }
+}
 
-
+void MDCache::purge_inodes(const interval_set<inodeno_t>& inos, LogSegment *ls)
+{
+  auto cb = new LambdaContext([this, inos, ls](int r){
+      assert(r == 0 || r == -2);
+      bool locked = false;
+      if (mds->mds_lock.is_locked_by_me())
+	locked = true;
+      
+      if (!locked)
+	mds->mds_lock.lock();
+      mds->inotable->project_release_ids(inos);
+      version_t piv = mds->inotable->get_projected_version();
+      assert(piv != 0);
+      mds->mdlog->start_submit_entry(new EPurged(inos, piv, ls->seq),
+				     new C_MDS_purge_completed_finish(this, inos, piv, ls));
+      mds->mdlog->flush();
+      if (!locked)
+	mds->mds_lock.unlock();
+    });
+  
+  dout(10) << __func__ << " start purge data : " << inos << dendl;
+  C_GatherBuilder gather(g_ceph_context);
+  SnapContext nullsnapc;
+  Filer filer(mds->objecter, mds->finisher);
```

Comment:
This Filer is destructed at the end of this method which should cause all sorts of havoc in the Filer callbacks? Why not use `MDCache::filer`?

Reply:
sorry. I didn't find that this pointer was used in the method before. 
There was no error at runtime,the reason may be when the service was started, there was less use of no memory. Although the destructor destroyed the filer class, the memory of object was not cleared, so it can still be used. It has been modified.
the error happened when make check is related to readable.sh.
what can i do to fix it? 
thx

Reply:
Probably unrelated; I'm retesting.

Reply:
Please rebase.

Reply:
I have rebased it now. thx